Notification scripts update

erikyo opened the pull request:
Since the previous version has all hub notifications hardcoded and the hub could not be tested, I created a react component and a redux store controller to handle notifications multiple destinations. The component has a context (formerly called location), which subscribes to notifications from its context and receives updates from it.

Then `notifyController` was rebuilt with all the previous functionality (add, remove, clearAll) and, in addition, a function was added to retrieve the api (it currently reads a false response from a json file).

Actually we have two main components: `Notification` and `NotificationArea`, the latter depending on context displays the array of `Notification` with diffent styles and options (for example at the moment the hub is set to display two lists of notifications, the current week's and previous week notices. [This is not yet customizable](https://github.com/erikyo/wp-notify/blob/443873fc9e93cdbe83e92df64ff1708de85b4b4f/src/scripts/components/notifications.js#L72-L80) but it would be nice if we can provide some filters (by type, by plugin etc as mentioned in PR https://github.com/WordPress/wp-feature-notifications/issues/84)

We have a rest-api endpoint (thanks @Sephsekla!) but we need to decide how the component fetches data from the database, what arguments it passes to it, and the strategy for updating the data in the reducer.

a brief story of what is changed with this PR:
- Adopted the `@wordpress/components` in order to comply to wordpress design standards
- Adopted the `@wordpress/data` to dispatch notifications to all components that request them
- The old function `addNotify` has been added to window.wp.notify (the js/api was located there)
https://github.com/erikyo/wp-notify/blob/443873fc9e93cdbe83e92df64ff1708de85b4b4f/src/scripts/wp-notify.js#L50
- When you test this PR remember to update the timestamp in the json file and this way moment.js will tell you how much time has passed in a textual way ("an hour ago", "few second ago", "a week ago" and so on)
- Adds the storybook and an initial component setup
- Fixes some a11y related issues (mainly now the hub is operable via keyboard)
- UI: the vertical padding of dash notifications should be a little smaller now on horizontal monitors with low vertical resolutions, because now it depends on the height of the screen
- Settings table updated (#110)

erikyo committed:
js import cleanup

erikyo committed:
move addNotify function into window.wp.notify (as the other wordpress js objects)

erikyo committed:
if the notification is empty (title and description) display a warning message in console

erikyo committed:
Notification management with redux
Better code splitting

erikyo committed:
adds a fetch function

erikyo committed:
Notification component update in order to support notification hub format
replaces the hardcoded notifications

erikyo committed:
Notification Hub - split by date the retrieved notifications

erikyo committed:
adds the state "status" used to control the appearance of the notification (fadeOff animation included)

erikyo committed:
adds @wordpress/components to comply WordPress standards

erikyo committed:
init storybook

erikyo committed:
Loads the basic Wordpress style into storybook
Enqueue the notify script after wp-components and wp-i18n has been loaded

erikyo committed:
add the jsDoc
updates dependencies
rearrange code

erikyo committed:
move single notice content into a wrapper in order to being able to nimbly style the border since it would now be a bit complex to do
added more jsdocs

erikyo committed:
notification message should have a visible focus state (close #108)
not 100% sure if it is wai compliant
we need to create the shortcut for action (space?) / delete (del ?) etc.
https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_roving_tabindex

erikyo committed:
The images associated with notifications should be considered decorative, but need an empty alt attribute to prevent screen readers from attempting to interpret them. (addresses #107)

erikyo committed:
wpNotifyHub styles
focus / active single hub notification

erikyo committed:
Accessibility: focus state on 'notifications settings' marginal. close #109

erikyo committed:
? 😓

erikyo committed:
A better way of assigning classes that does not create white space

erikyo committed:
chore

erikyo committed:
adds wordpress docgen script

erikyo committed:
storybook 6.5 configuration

erikyo committed:
adds the WordPress backend assets to storybook

erikyo committed:
try to handle vertical padding for smaller screens

Sephsekla commented on includes/demo.php:
```diff
@@ -16,11 +16,10 @@ function wp_admin_bar_wp_notify_item( $wp_admin_bar ) {
 
 	$footer = '<footer>
-    <a>
       <a href="' . esc_url( admin_url( 'options-general.php?page=wp-notify' ) ) . '" class="wp-notification-action wp-notification-action-markread button-link">
```
Not important since this is just demo code, but in future I'd probably use `printf` for these in order to make things a bit easier to read.

Sephsekla commented on package.json:
```diff
@@ -2,17 +2,17 @@
+  "author": "The WordPress Contributors",
+  "license": "GPL-2.0-or-later",
+  "bugs": {
+    "url": "https://github.com/WordPress/wp-notify/issues"
```
```suggestion
    "url": "https://github.com/WordPress/wp-feature-notifications/issues"
```

While we're in here

Sephsekla commented on package.json:
```diff
@@ -2,17 +2,17 @@
+  "bugs": {
+    "url": "https://github.com/WordPress/wp-notify/issues"
+  },
+  "homepage": "https://github.com/WordPress/wp-notify#readme",
```
```suggestion
  "homepage": "https://github.com/WordPress/wp-feature-notifications#readme",
```

Sephsekla commented on package.json:
```diff
@@ -34,22 +34,47 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/WordPress/wp-notify.git"
```
```suggestion
    "url": "git+https://github.com/WordPress/wp-feature-notifications.git"
```

Sephsekla commented on src/scripts/components/noticeImage.js:
```diff
@@ -0,0 +1,25 @@
+
+import wpLogo from '../../images/WordPressLogo.svg';
+
+export default class NoticeImage extends Component {
```
I have a few thoughts on this, but most can wait for a followup later.

Each of these components could do with a comment though.

Sephsekla commented on src/scripts/components/notification.js:
```diff
@@ -0,0 +1,126 @@
+	render() {
+		const classes = [
+			'wp-notification',
+			'wp-notice-' + this.props.id,
```
As a general note, I personally tend to prefer something like

```js
const { additional, dismissible, id /* ..etc */ } = this.props
```

up top, and then you can just refer to `id` or `dismissable` in the rest of the code.

Sephsekla commented on src/scripts/components/notification.js:
```diff
@@ -0,0 +1,126 @@
+import { delay } from '../utils';
+
+// Notice class method.
+export default class Notice extends Component {
```
Does this have to be a class? Or could it just be `const` or `function`? It's not got any class-specific logic or anything.

erikyo replied:
For certain things I have always preferred to create a class whereas I use functional components more often for component elements, this one has also a method (onDismiss) and I guess more may be added in the future..
In this way the component is also more prone to being extended (not that it would be impossible otherwise)

Sephsekla commented on src/scripts/components/notification.js:
```diff
@@ -0,0 +1,126 @@
+			this.props.severity || null,
+			this.props.unread ? 'unread' : null,
+			this.props.additionalClassName,
+		].filter( Boolean );
```
Have you come across https://github.com/JedWatson/classnames? This will do a lot of the heavy lifting here for you.

It's already in use in Gutenberg, so there's a precedent.

Sephsekla commented on src/scripts/components/notification.js:
```diff
@@ -0,0 +1,126 @@
+			this.props.additionalClassName,
+		].filter( Boolean );
+
+		return (
```
This is good, but it is a bit of a massive thing 😅 

Maybe not now, but it would be good to split it up slightly, so something like this for each of the individual conditional components.

```js

const dashboardActions = () => <something>Some buttons</something>

return (
  <div>
    { (location === 'dashboard') && dashboardActions }
  </div>
)

erikyo replied:
Following this idea I also created a component for the actions (ok+dismiss button), but I was wondering if it would be a good idea to do the same for the title/text. What do you think?

Sephsekla commented on src/styles/notify/notification-hub.scss:
```diff
@@ -24,6 +24,11 @@
   }
 
+  // reset child icons color since by default the icon is lightened when the mouse is in hover state
+  #wpadminbar &:hover #wp-notify-hub .ab-icon:before {
```
These hover states should probably apply to `:focus` as well, since keyboard nav should have the same experience

Sephsekla commented on src/styles/notify/notification-hub.scss:
```diff
@@ -24,6 +24,11 @@
     transform: translateX(-100%);
```
I'd like to split some of these files up a bit further in an ideal world.

Sephsekla reviewed (commented):
@erikyo thanks for all your hard work here! I've added a few comments but in general this has massively pushed things forward. 

As a general note I'd like to do some work on modularisation and comments in particular, but I think it would make more sense to get this merged and then we can chip away at any updates in smaller PRs (since this is pretty huge at this point).

@bacoords @jonathanbossenger what are your thoughts? Should we go ahead with a merge so we can work against `develop` again?

erikyo committed:
Update package.json

Co-authored-by: Joe Bailey-Roberts <33831317+Sephsekla@users.noreply.github.com>

erikyo committed:
Update package.json

Co-authored-by: Joe Bailey-Roberts <33831317+Sephsekla@users.noreply.github.com>

erikyo committed:
Update package.json

Co-authored-by: Joe Bailey-Roberts <33831317+Sephsekla@users.noreply.github.com>

erikyo committed:
scripts and styles enhancements:
- better code splitting
- reworked webpack configuration
- adds dompurify and propTypes checks

erikyo committed:
remove unused import
reverts "Notifications"

erikyo committed:
handles empty drawer

erikyo committed:
fix for metabox that doesn't clear notifications

erikyo committed:
closes the drawer with the "esc" key (ref #111)

erikyo committed:
updates the js api - add list and find
List - It returns the notifications from the Redux store a search term
Find - It searches the Redux store for a notification by ID or by a search term

erikyo committed:
date fix (we need epoch in seconds and not in milliseconds)

erikyo committed:
provides a frontend js api access for the list and the find requests

Sephsekla reviewed (commented):
👍 We can definitely iterate on this, but for now (once conflicts are resolved) I'd be in favour of getting this merged so we can move to smaller PRs against `develop` (since a lot of the progress is still only in this branch.

erikyo committed:
remove unused import
reverts "Notifications"

erikyo committed:
renames NoticeImage component

erikyo committed:
updated components (Notice & NoticeArea)

erikyo committed:
switching from redux-toolbox to wordpress/data store

erikyo committed:
updated plugin init and notices area initialization

erikyo committed:
adds the id to each notification
location renamed into context
lint

erikyo committed:
repo configuration

erikyo committed:
updated Notice component

erikyo committed:
the two components image and icon were combined

erikyo committed:
adds a sorting function used to split notices

erikyo committed:
adds findNotice function

erikyo committed:
findContext is a new function that walk through the state object and looks for the notification with the given id

erikyo committed:
adds clear to reducer
updated reducer functions

erikyo committed:
synchronizes init with new reducer changes

erikyo committed:
adds constants.js to store
adds some missing comments

erikyo committed:
Notice action component (hold buttons and similar stuff)

erikyo committed:
The text and icon shown when there are no notices in the drawer

erikyo committed:
the Component that holds the loop of notices and its wrapper

erikyo committed:
the drawer section panel header

erikyo committed:
Provides the ability to interact with the store using the metabox buttons

erikyo committed:
updated fake api response

erikyo committed:
Updating the style to make it more accessible (and also more readable)
closes #109

erikyo committed:
Replaces the settings table with one made in the wordpress way
closes #110

erikyo committed:
makes the parts of the dashboard work in accordance with the plugin.
this is somehow related to #110 because it makes the drawer open when selected with tab

erikyo committed:
updates components to use the new style of the previous commit
fix for an action dispatch that was called incorrectly

erikyo committed:
Adds husky (that replaces the previous git hooks handled with composer)

erikyo commented:
since [989c918](https://github.com/WordPress/wp-feature-notifications/pull/92/commits/989c918b912c8acd5634abeb00ffc3274e9eb59f) checks are failing since i've removed the rules exclusions in the phpcs configuration file. 

I think if we want use the WP-standards we can't pick and choose the ones we like or don't like and that since it's only loaded the ruleset of core and doc we should follow it.

erikyo committed:
updates exclude pattern

Sephsekla commented:
> since [989c918](https://github.com/WordPress/wp-feature-notifications/pull/92/commits/989c918b912c8acd5634abeb00ffc3274e9eb59f) checks are failing since i've removed the rules exclusions in the phpcs configuration file.
> 
> I think if we want use the WP-standards we can't pick and choose the ones we like or don't like and that since it's only loaded the ruleset of core and doc we should follow it.

@erikyo for now can you just revert the phpcs config change? While I agree that it's a change worth making, I'd prefer to make it in its own followup PR. It's a little outside the scope of JS scripts update, and ideally something to handle separately rather than delaying merging this into develop.

erikyo committed:
reverts phpcs.xml.dist

erikyo committed:
reverts changes to demo (will be moved to another PR)

erikyo committed:
storybook: updated with newer components

erikyo committed:
adding a new component "NoticeMeta" that shows source and time
NoticeAction component updated with the standard WordPress css classes

erikyo committed:
Notices Hub header and footer components

erikyo committed:
lint/fixing jsdoc

erikyo committed:
updated drawer (with components and event)

erikyo committed:
updated style (to reflect changes in ./src/components)

Sephsekla reviewed (approved):
💥 Awesome stuff!

Re: the changes to the json file, I think this is fine for now, although I'd like to get the ball rolling again on the schema discussion and finalise this before we go much further.

There's a lot in here and probably a few things that could be tweaked, but for now I'd be in favour of merging this and iterating on things in smaller PRs going forward.

erikyo committed:
Merge branch 'develop' into js-develop

Sephsekla reviewed (approved):
🥳
